### DISCUSSION TITLE
Does `abitype` work on JSON files imported as modules?

### DISCUSSION THREAD

**Discussion by typedarray** [wagmi-dev/abitype#46] [Q&A] [answered]
Hi! Thanks for creating `wagmi` and `abitype`! 👑 👑

I'm developing a [framework](https://github.com/0xOlias/ponder) that currently uses TypeChain to generate contract types for user-provided ABIs. I'd much rather use this library, but I'm having trouble using `abitype` on JSON files imported as modules. [This repo](https://github.com/0xOlias/abitype-example) minimally demonstrates the issue.

I followed some IntelliSense suggestions to add the following to `tsconfig.json`:
```ts
// tsconfig.json
"resolveJsonModule": true,
"allowSyntheticDefaultImports": true
```

But I still get the following TypeScript errors:

<img width="639" alt="Screen Shot 2022-10-20 at 5 14 18 PM" src="https://user-images.githubusercontent.com/90073088/197060485-379c225b-18bd-4ecc-8bfb-90d892fdf357.png">

I'm a relative TypeScript noob so I could be missing something fundamental here. Thanks!

**Comment by tmm** [comment, marked as answer]
Unfortunately [TypeScript doesn't support importing JSON as const](https://github.com/microsoft/TypeScript/issues/32063) right now. The best option is to change the file to `.ts` and add a const assertion there.

**Reply by roninjin10** [reply]
This can cause pretty disastrous things to happen to total typecheck performance of not implemented smartly.   You would want to make as const explicit for this reason so I wouldn't expect ts to implement this anytime soon

**Reply by typedarray** [reply]
@roninjin10 do you know if it would be possible to write a ts lang service plugin that does this, but only for files with a specific suffix like `.abi.json`? That seems like it would limit the blast radius.

**Reply by roninjin10** [reply]
Yea definitely possible.  But one caveat is that ts service language plugins only get loaded by the language server in your editor (vscode, jetbrains, coc ect.). They do not get loaded when you run `tsc` in  the terminal.  I have a of paths to explore on how to solve this:

- provide an alternative to tsc to run `npx @evmts/ts-plugin tsc`
- generate a .d.ts that then can be used by tsc when typechecker runs `yarn @evmts/ts-plugin generate && tsc`
- See how other tools with custom ts plugins like astro handle it

But basically because TS explicitly doesn't support ts lang service plugins with tsc getting the plugin working is the easy part, getting it working in a way that doesn't break other tools is the hard part.

**Reply by tmm** [reply]
It's pretty straightforward to [write a plugin](https://github.com/microsoft/TypeScript/wiki/Writing-a-Language-Service-Plugin#overview-writing-a-simple-plugin), but language service plugins only improve the editing experience and cannot customize the type system to change what is or isn't an error when running `tsc`.

For example, the follow would show an error in your editor `Type '"foo"' is not assignable to type '"balanceOf" | ... '.`, but wouldn't fail as expected when compiling.

```ts
import erc20Abi from './abi.json'

type Result = ExtractAbiFunctionNames<typeof erc20Abi>
const functionName: Result = 'foo'
```

**Reply by roninjin10** [reply]
Yea for EVMts this issue is not a negative tradeoff because EVMts already requires a bundler to compile.   As long as I add back a way of doing typechecking too it's not an issue.

As a general solution though it's a bad approach.    Not supporting `tsc` as a compiler is a pretty huge tradeoff.

**Reply by typedarray** [reply]
Understood, thanks for the thorough answers.

**Reply by roninjin10** [reply]
@0xOlias  One easy way to fix your issue that works as long as all your abis all have unique names is to generate a `global.d.ts` file

```typescript
declare module '*/SomeAbi.json' {
  export default [
        {
          "inputs": [
          ....
          ]
        }
  ] as const
}
```

This is so far my best answer for how to get typechecking working in evmts so far but hoping to find a solution that doesn't have that limitation of every file having a unique name

**Comment by razgraf** [comment]
For anyone else in need it looks like @tmm said, `.ts` imports are still the way to go. There's an extra catch though: make sure to have the following structure in those files:

```ts
const abi = [...] as const;
export ...
```

If you import the object from the `.ts` file and cast it to `const` afterwards, Typescript might scream at you. The cast needs to happen inline for it to work properly.

**Comment by startswithaj** [comment]
Is this possible today? Without rename the ABI's to `.ts`?
